fruitCollection.js

```diff
@@ -0,0 +1,115 @@
+/*
+monkey eating fruit
+tree to tree
+find collection of tree of fruit
+
+list of tree
+each tree: x-coordinate, y-coordinate, num of fruit, vine length
+
+i: [2,4,5,10]
+o: max of fruit 
+
+const trees = [
+    [2, 4, 2, 6],
+    [10, 4, 5, 5],
+    [3, 8, 3, 6],
+    [12, 8, 6, 5],
+    ,[1, 6 2, 6],
+];
+
+*/
+
+const getDistanceBetweenNodes = (nodeOneCoordinates, nodeTwoCoordinates) => {
+    const [nodeOneXcoordinate, nodeOneYCoordinate] = nodeOneCoordinates;
+    const [nodeTwoXcoordinate, nodeTwoYCoordinate] = nodeTwoCoordinates;
+
+    const distance = Math.sqrt(
+        Math.pow(nodeTwoXcoordinate - nodeOneXcoordinate, 2) +
+            Math.pow(nodeTwoYCoordinate - nodeOneYCoordinate, 2),
+    );
+
+    return distance;
+};
+
+const areConnected = (nodeOne, nodeTwo) => {
+    const [nodeOneXcoordinate, nodeOneYCoordinate, nodeOneFruitNum, nodeOneLengthDistance] =
+        nodeOne;
+    const [nodeTwoXcoordinate, nodeTwoYCoordinate, nodeTwoFruitNum, nodeTwoLengthDistance] =
+        nodeTwo;
+
+    const distanceBetweenNodes = getDistanceBetweenNodes(
+        [nodeOneXcoordinate, nodeOneYCoordinate],
+        [nodeTwoXcoordinate, nodeTwoYCoordinate],
+    );
+
+    const smallestDistanceBetweenTwoNodes = Math.min(nodeOneLengthDistance, nodeTwoLengthDistance);
+    return distanceBetweenNodes <= smallestDistanceBetweenTwoNodes;
+};
+
+const buildGraph = (nodes) => {
+    const graph = {};
+    const numNodes = nodes.length;
+
+    for (let nodeOne = 0; nodeOne < numNodes; nodeOne++) {
+        for (let nodeTwo = nodeOne + 1; nodeTwo < numNodes; nodeTwo++) {
+            if (!areConnected(nodes[nodeOne], nodes[nodeTwo])) continue;
+
+            const containsNodeOne = graph.hasOwnProperty(nodeOne);
+            const containsNodeTwo = graph.hasOwnProperty(nodeTwo);
+
+            if (!containsNodeOne) graph[nodeOne] = [];
+            if (!containsNodeTwo) graph[nodeTwo] = [];
+
+            graph[nodeOne].push(nodeTwo);
+            graph[nodeTwo].push(nodeOne);
+        }
+    }
+
+    return graph;
+};
+
+const getComponent = (startNode, graph, visited, nodes) => {
+    const queue = new Queue();
+
+    visited.add(startNode);
+    queue.enqueue(startNode);
+
+    let numAmount = 0;
+    while (queue.size() > 0) {
+        //  Remove node
+        const node = queue.dequeue();
+
+        // Process node
+        numAmount += nodes[node][2];
+
+        // Get neighbors
+        const containsNode = graph.hasOwnProperty(node);
+        if (!containsNode) continue;
+
+        const neighbors = graph[node];
+        for (const neighbor of neighbors) {
+            if (visited.has(neighbor)) continue;
+
+            visited.add(neighbor);
+            queue.enqueue(neighbor);
+        }
+    }
+
+    return numAmount;
+};
+
+const getMaxFruit = (trees) => {
```

Comment:
P: Why are you using a BFS and not a DFS?

Reply:
Since we're trying to find the maximum component size (in this case fruit), it would be faster to use a BFS than a DFS because we can traverse in levels and also, if we're using BFS, we're guaranteed the shortest path in doing so.

Reply:
Yes. To add a bit more to this, our goal here is the explore the ENTIRE component. So, it doesn't matter which order we traverse these nodes in. In something like a shortest path problem, we are NOT necessarily going to traverse the whole component, so we need to be smart about the order in which we traverse the nodes.